src/react/ReactPropsSet.js

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { Realm } from "../realm.js";
+import { ObjectValue, Value } from "../values/index.js";
+import invariant from "../invariant.js";
+import { type ReactSetKeyMap, ReactSet } from "./ReactSet.js";
+import { ResidualReactElementVisitor } from "../serializer/ResidualReactElementVisitor.js";
+
+export class ReactPropsSet extends ReactSet {
+  constructor(realm: Realm, residualReactElementVisitor: ResidualReactElementVisitor) {
+    super(realm, residualReactElementVisitor);
+    this.reactPropsRoot = new Map();
+  }
+  reactPropsRoot: ReactSetKeyMap;
+
+  add(props: ObjectValue, visitedValues: Set<Value> | void): ObjectValue {
+    if (!visitedValues) visitedValues = new Set();
+    let currentMap = this.reactPropsRoot;
+    let result;
+
+    for (let [propName] of props.properties) {
+      currentMap = this._getKey(propName, currentMap, visitedValues);
+      let prop = this._getEquivalentPropertyValue(props, propName);
+      result = this._getValue(prop, currentMap, visitedValues);
+      currentMap = result.map;
+    }
+    let temporalAlias = props.temporalAlias;
+
+    if (temporalAlias !== undefined) {
+      currentMap = this._getKey("temporalAlias", currentMap, visitedValues);
+      result = this._getValue(temporalAlias, currentMap, visitedValues);
```

Comment:
What case does this handle?

I think this would break if there was a prop called `temporalAlias` (very unlikely but should be easy to fix).

Reply:
It's for snapshotting, I've addressed in latest commit :)

Reply:
Can we use a symbol or an object as a key in this case? String clashes with user supplied props (which also use strings).